### PR TITLE
Change the way editor might be passed to HotColumn and HotTable to use render props approach

### PR DESCRIPTION
**Pull request by NOtherDev**
### Context

Specifying editors as children components marked with "magic" `hot-editor` attribute does not feel idiomatic. It doesn't represent something that is a part of the column, it's purely the column's property. Moreover, this kind of composition allows putting multiple Editors in one column, which is illogical. 

Another issue is that a custom component needs to inherit from a specific base class, thus needs to be a class-based component, which isn't seen correct in modern React and might be hard to follow for some developers.

Instead, we're proposing to use a variant of render props pattern and allow specifying functional component-based editors via `HotTable` or `HotColumn` props directly. 

#### Implementation notes

Well, good luck to anyone reviewing this beast.

It took me quite a lot of time to get it to the working state and I'm not extremely satisfied with the shape it takes now, to the extent that I have some doubts if it's better than the previous approach. 

There are three layers of the solution: `HotTable` component and its `editor` prop, custom editor component and the dynamic class that bridges them in order to satisfy native Handsontable API. In the previous approach the editor component was required to be class-based, inheriting from a provided base class. In the new approach it needed to be functional, so bridging it to the native Handsontable API via dynamic class is more challenging.

In [the discussion in #3](https://github.com/NOtherDev/handsontable/pull/3#issuecomment-1933606423) we figured out that `useImperativeHandle` might be a good fit to provide HOT editor hook functions from the inner custom editor component to the outer code (eventually to the bridging class), although during the implementation I discovered a bit more constraints/consequences of this approach:
* HOT editor hook functions sometimes need to reference other functions or call base (super class) behaviors; while others might be called with properly bound `this`, there's no way to emulate `super` pointer easily in non-class-based code
* there are use cases where the HOT-native editor instance needs to be accessed from within the custom editor component, while it isn't created yet at the moment the custom editor is mounted, so `ref` access was needed; it means there are two-way refs: custom editor sends its hook functions up via context and `useImperativeHandle` while it receives back the HOT-native editor instance back via another ref.
* I ended up creating a helper hook `useHotEditor` that wraps the above - it hides `useImperativeHandle`, provides `super` emulation and returns editor instance ref

These together make the final user-facing API like this:

```
<HotTable editor={CustomEditor} ... />
```
or, if it needs to be customized: 
```
<HotTable editor={() => <CustomEditor customProp={1} />} ... />
```

+ the editor itself:
```
const CustomEditor = (props: Props) => {
  const hotEditorInstanceRef = useHotEditor(ref, (runSuper) => ({
    prepare(...args) {
      runSuper().prepare(...args); // used to be super.prepare(...)
      this.otherHook(); // calls BaseEditor method
      // whatever custom code
    }
  });

 const finish = () => hotEditorInstanceRef.current.finishEditing(); // used to be this.finishEditing()

  return (
    <button onClick={finish}>test</>
  );
};
```

### How has this been tested?

* All the unit tests still pass + new behavior tests implemented.
* Example projects from the repo still work properly.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. Discussion on RFC: https://github.com/handsontable/handsontable/discussions/10767
2. Discussion on RFC draft: https://gist.github.com/evanSe/8c67315d837873c7f32f7446f027e5e7
3. #3 

### Affected project(s):
- [ ] `handsontable`
- [ ] `@handsontable/angular`
- [x] `@handsontable/react`
- [ ] `@handsontable/vue`
- [ ] `@handsontable/vue3`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [x] My change requires a change to the documentation.

### CODE REVIEW

File: wrappers/react/test/componentInternals.spec.tsx [hunk 1]
```diff
@@ -1,146 +1,16 @@
...
   mockElementDimensions,
   mountComponentWithRef,
+  renderComponentWithProps,
   sleep,
 } from './_helpers';
 import { HOT_DESTROYED_WARNING } from "../src/helpers";
-import { BaseEditorComponent } from '../src/baseEditorComponent';
-import { HotRendererProps } from '../src'
-
-describe('Subcomponent state', () => {
```
Comment (NOtherDev) [inline wrappers/react/test/componentInternals.spec.tsx:15]: In function-based components manipulating state from the outside isn't possible, so these tests are no longer relevant

**Comment by NOtherDev** [comment]
@evanSe 
Please have a look at the description above - I tried to summarize the struggle here. I'm open to have a call to go through the complexities of this solution and I'm especially open for suggestions how can we simplify. Or even I'm open to roll back what was decided before and try another approach...

Asking for your review already, even though there are examples and docs changes missing + proper validation. I'll follow up on these once the solution is validated.

**Comment by evanSe** [comment]
> @evanSe Please have a look at the description above - I tried to summarize the struggle here. I'm open to have a call to go through the complexities of this solution and I'm especially open for suggestions how can we simplify. Or even I'm open to roll back what was decided before and try another approach...
> 
> Asking for your review already, even though there are examples and docs changes missing + proper validation. I'll follow up on these once the solution is validated.

I have had some time to go through this today and I think we can make it work! its a beast to review 😄 but I think with a bit of love around the API surface layer this can be perfect for users.

I will try find some time soon for suggestions on how to make it the best DX for end user.

**Comment by evanSe** [comment]
I updated a rather ugly component we had around to this branch and tried it out, it worked great except for `useState` not working in the `getValue` part as I think it is destroyed by the time getValue is called

```tsx
import { useRef, useState } from "react";
import { HexColorPicker } from "react-colorful";
import { hotEditor, useHotEditorHooks } from "../wrapper";

const ColorPickerEditor = hotEditor((props, ref) => {
  const [color, setColor] = useState('');
  const mainElementRef = useRef<HTMLDivElement>(null)

  const hotCustomEditorInstanceRef = useHotEditorHooks(ref, (superBoundEditorInstanceProvider) => ({
    getValue() {
      console.log('getValue', color)
      return color;
    },

    setValue(value, ...rest) {
      console.log('setValue', {value, rest})
      setColor(value)
    },

    prepare(row, col, prop, TD, originalValue, cellProperties): any {
      superBoundEditorInstanceProvider().prepare(row, col, prop, TD, originalValue, cellProperties)
      const tdPosition = TD.getBoundingClientRect();

      if (!mainElementRef.current) return;

      mainElementRef.current.style.left =
        tdPosition.left + window.pageXOffset + "px";

      mainElementRef.current.style.top =
        tdPosition.top + window.pageYOffset + "px";
    },

    open() {
      if (mainElementRef.current) mainElementRef.current.style.display = 'block';
      (
        document.querySelector(".react-colorful__interactive") as HTMLDivElement
      )?.focus();
    },

    close() {
      if (mainElementRef.current) mainElementRef.current.style.display = 'none';
    }
  }), [color, setColor, mainElementRef]);

  const handleMouseDown = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    e.stopPropagation();
  }
  const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
    e.stopPropagation();
    if (e.key === "Escape") {
      hotCustomEditorInstanceRef?.current?.finishEditing()
    }
  }

  return (
    <div
      ref={mainElementRef}
      style={{
        display: "none",
        position: "absolute",
        left: 0,
        top: 0,
        zIndex: 999,
        background: "#fff",
        padding: "15px",
        border: "1px solid #cecece",
      }}
      onMouseDown={handleMouseDown}
      onKeyDown={handleKeyDown}
    >
      <HexColorPicker
        color={color}
        onChange={(newColor) => setColor(newColor)}
      />
      <button
        style={{ width: "100%", height: "33px", marginTop: "10px" }}
        onClick={() => hotCustomEditorInstanceRef?.current?.finishEditing()}
      >
        Apply
      </button>
    </div>
  );
});

export default ColorPickerEditor;
```

I will find time over the next two days to make suggestions about the API we expose to the end user, but I think you should be fine to proceed with any refactors you think are needed.

**Comment by NOtherDev** [comment]
> I updated a rather ugly component we had around to this branch and tried it out, it worked great except for `useState` not working in the `getValue` part as I think it is destroyed by the time getValue is called

Good catch, fixed now 🙂

File: wrappers/react/src/hotEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+/**
+ * Hook that allows encapsulating custom behaviours of component-based editor by customizing passed ref with overridden hooks object.
+ *
+ * @param {React.Ref} ref Reference for component-based editor overridden hooks object.
+ * @param {Function} overriddenHooks Function that provides the overrides specific for the custom editor.
+ *  It gets an object that emulates super calls to BaseEditor methods, if needed.
+ * @param {React.DependencyList} deps Overridden hooks object React dependency list.
+ * @returns {React.RefObject} Reference to Handsontable-native editor instance.
+ */
+export function useHotEditorHooks(ref: React.Ref<HotEditorHooks>, overriddenHooks?: (superBoundEditorInstanceProvider: () => Handsontable.editors.BaseEditor) => HotEditorHooks, deps?: React.DependencyList): React.RefObject<Handsontable.editors.BaseEditor> {
```
Comment (evanSe) [inline wrappers/react/src/hotEditor.tsx]: Lets rename this to `useHotEditor`

File: wrappers/react/src/hotEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    {children}
+  </EditorContext.Provider>
+}
+
+/**
+ * Helper function that wraps React.ForwardRef for ease of creating custom component-based editors.
+ *
+ * @param {Function} render The editor component function.
+ */
+export function hotEditor<P>(render: React.ForwardRefRenderFunction<HotEditorHooks, P>) {
```
Comment (evanSe) [inline wrappers/react/src/hotEditor.tsx]: lets remove this helper function that wraps all together, its another level of friction from the end user we want to avoid.

File: wrappers/react/src/hotEditor.tsx (hunk 2 above)
Comment (evanSe) [reply wrappers/react/src/hotEditor.tsx]: I also propose we pass the hooks reference via context rather than as a parameter to `useHotEditorHooks`

example code I had working
```tsx
export function useHotEditor(overriddenHooks?: (superBoundEditorInstanceProvider: () => Handsontable.editors.BaseEditor) => HotEditorHooks, deps?: React.DependencyList): React.RefObject<Handsontable.editors.BaseEditor> {
// example name, a better one probably exists
  const hooksRef = useContext(HooksContext);
  const hotCustomEditorInstanceRef = React.useContext(EditorContext);
  const superBoundEditorInstanceProvider = () => superBound(hotCustomEditorInstanceRef.current);
  React.useImperativeHandle(hooksRef, () => overriddenHooks?.(superBoundEditorInstanceProvider) || {}, deps);
  return hotCustomEditorInstanceRef;
}

```

File: wrappers/react/src/hotEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  return React.forwardRef<HotEditorHooks, P>(render);
+}
+
+/**
+ * Hook that allows encapsulating custom behaviours of component-based editor by customizing passed ref with overridden hooks object.
+ *
+ * @param {React.Ref} ref Reference for component-based editor overridden hooks object.
```
Comment (evanSe) [inline wrappers/react/src/hotEditor.tsx]: this would be removed and we could use context

File: wrappers/react/src/helpers.tsx [hunk 5]
```diff
@@ -48,66 +45,54 @@ export function warn(...args) {
...
+  if (typeof editorComponent === 'function') {
+    Editor = React.forwardRef(editorComponent as React.ForwardRefRenderFunction<HotEditorHooks>);
+  } else {
+    Editor = editorComponent as React.ForwardRefExoticComponent<React.RefAttributes<HotEditorHooks>>;
+  }
+
+  const editorElement = <Editor ref={hooksReference} />;
   const containerProps = getContainerAttributesProps(editorElement.props, false);
 
   containerProps.className = `${DEFAULT_CLASSNAME} ${containerProps.className}`;
```
Comment (evanSe) [inline wrappers/react/src/helpers.tsx:89]: I suggest replacing the following code
```ts
  return ReactDOM.createPortal(
    <div {...containerProps}>
      {editorElement}
    </div>
    , doc.body);
```
with
```ts
  return ReactDOM.createPortal(
    <div {...containerProps}>
      <HooksContext.Provider value={hooksReference}>
        {editorElement}
      </HooksContext.Provider>
    </div>
    , doc.body);
```

File: wrappers/react/src/helpers.tsx [hunk 6]
```diff
@@ -48,66 +45,54 @@ export function warn(...args) {
...
 
+  let Editor: React.ForwardRefExoticComponent<React.RefAttributes<HotEditorHooks>>;
+
+  if (typeof editorComponent === 'function') {
+    Editor = React.forwardRef(editorComponent as React.ForwardRefRenderFunction<HotEditorHooks>);
+  } else {
+    Editor = editorComponent as React.ForwardRefExoticComponent<React.RefAttributes<HotEditorHooks>>;
+  }
+
+  const editorElement = <Editor ref={hooksReference} />;
```
Comment (evanSe) [inline wrappers/react/src/helpers.tsx]: with the suggestion in my other comment, we can then avoid passing the ref to the Editor altogether as it will come via context
```suggestion
  const editorElement = <Editor  />;
```

**Review by evanSe — CHANGES_REQUESTED** [review]
Hey I made some suggestions to simplify the end users experience, namely getting rid of the `hotEditor` forward ref and passing this "hooksRef" via context instead which I tested and and works!

here is an example improved experience for the user after those changes with a ColorEditor
```tsx
const ColorEditor = () => {
  const [value, setValue] = useState("");
  const mainElementRef = useRef<HTMLDivElement>(null);

  const table = useHotEditor(
    (runSuper) => ({
      getValue() {
        return value;
      },
      setValue(value, ...rest) {
        setValue(value);
      },
      prepare(row, col, prop, TD, originalValue, cellProperties): any {
        runSuper().prepare(
          row,
          col,
          prop,
          TD,
          originalValue,
          cellProperties
        );
        const tdPosition = TD.getBoundingClientRect();

        if (!mainElementRef.current) return;

        mainElementRef.current.style.left =
          tdPosition.left + window.pageXOffset + "px";

        mainElementRef.current.style.top =
          tdPosition.top + window.pageYOffset + "px";
      },

      open() {
        if (mainElementRef.current)
          mainElementRef.current.style.display = "block";
      },

      close() {
        if (mainElementRef.current)
          mainElementRef.current.style.display = "none";
      },
    }),
    [value, setValue, mainElementRef]
  );

  

  const handleMouseDown = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    e.stopPropagation();
  };
  const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
    e.stopPropagation();
    if (e.key === "Escape") {
      table?.current?.finishEditing();
    }
  };

  return (
    <div
      ref={mainElementRef}
      style={{
        display: "none",
        position: "absolute",
        left: 0,
        top: 0,
        zIndex: 999,
        background: "#fff",
        padding: "15px",
        border: "1px solid #cecece",
      }}
      onMouseDown={handleMouseDown}
      onKeyDown={handleKeyDown}
    >
      <input
        value={value}
        type="text"
        onChange={(event) => setValue(event.currentTarget.value)}
      />
      <button
        style={{ width: "100%", height: "33px", marginTop: "10px" }}
        onClick={() => table?.current?.finishEditing()}
      >
        Apply
      </button>
    </div>
  );
};
```

However I think we could manage to do some further abstraction and achieve this
```tsx
const ColorEditor = () => {
  const mainElementRef = useRef<HTMLDivElement>(null);

  const table = useHotEditor({
    // perhaps this could be hiden from the user?
    prepare(): any {
      // ommited for example
    },
    onOpen() {
      if (mainElementRef.current)
        mainElementRef.current.style.display = "block";
    },
    onClose() {
      if (mainElementRef.current) mainElementRef.current.style.display = "none";
    },
  });

  const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
    e.stopPropagation();
    if (e.key === "Escape") {
      table.close();
    }
  };

  return (
    <div
      ref={mainElementRef}
      style={{
        // ommited for example
      }}
      onKeyDown={handleKeyDown}
    >
      <input
        value={table.value}
        type="text"
        onChange={(event) => table.setValue(event.currentTarget.value)}
      />
      <button
        style={{ width: "100%", height: "33px", marginTop: "10px" }}
        onClick={() => table.close()}
      >
        Apply
      </button>
    </div>
  );
};
```

I have not settled on the final abstraction yet, so we should be good to move forward with merging this pull request once we pass the hooksRef via context.

On another note, I am trying to plan future releases if we ignore an improved abstraction how long to do you currently think it will take to finish everything up?

File: wrappers/react/src/helpers.tsx (hunk 5 above)
Comment (NOtherDev) [reply wrappers/react/src/helpers.tsx:89]: I've merged this `HooksContext` into `EditorContext` that I created before as they are actually running on the same level - works nice as well 🙂

**Comment by NOtherDev** [comment]
> Hey I made some suggestions to simplify the end users experience, namely getting rid of the `hotEditor` forward ref and passing this "hooksRef" via context instead which I tested and and works!

Perfect, it simplifies stuff a lot! Thx for noticing this! Updates pushed.

> However I think we could manage to do some further abstraction and achieve this
> 
> ```tsx
> const ColorEditor = () => {
>   const mainElementRef = useRef<HTMLDivElement>(null);
> 
>   const table = useHotEditor({
>     // perhaps this could be hiden from the user?
>     prepare(): any {
>       // ommited for example
>     },
>     onOpen() {
>       if (mainElementRef.current)
>         mainElementRef.current.style.display = "block";
>     },
> [cut]

I'm not sure it will be practical, as we need to be able to call `super` or make sure this hooks object is fresh once the dependencies change. But sure, I'm happy to iterate deeper if you think it's worth it.

> On another note, I am trying to plan future releases if we ignore an improved abstraction how long to do you currently think it will take to finish everything up?

I think this one is/was probably the hardest. Next I'm going to tackle the actual rewrite into functional components, but while it will be large in terms of lines changed as well, I don't foresee much complexities. And the further ones are rather cosmetic in comparison to these two, so I think we should be well done in a month's time.

Now, as you merged the first PR into the origin, I'll recreate #2 and #4 there as well.

**Review by evanSe — APPROVED** [review]
I say we leave further iteration/abstractions off the table for now.

happy to merge this one!

**Comment by NOtherDev** [comment]
> > On another note, I am trying to plan future releases if we ignore an improved abstraction how long to do you currently think it will take to finish everything up?
> 
> I think this one is/was probably the hardest. Next I'm going to tackle the actual rewrite into functional components, but while it will be large in terms of lines changed as well, I don't foresee much complexities. And the further ones are rather cosmetic in comparison to these two, so I think we should be well done in a month's time.

@evanSe seems I underestimated our ability to merge things into the upstream 😊 I have pushed last required PR and the last thing to cover is the data immutability issue. I'm planning to approach it next week. And that's it then, I think?
